ssl/quic/quic_port.c

```diff
@@ -1329,13 +1347,79 @@ static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
         *scid = hdr->src_conn_id;
     }
 
+    /*
+     * Determine if we need to send a NEW_TOKEN frame
+     * If we validated a retry token, we should always
+     * send a NEW_TOKEN frame to the client
+     *
+     * If however, we validated a NEW_TOKEN, which may be
+     * reused multiple times, only send a NEW_TOKEN frame
+     * if the existing received token has less than 10% of its lifetime
+     * remaining.  This prevents us from constantly sending
+     * NEW_TOKEN frames on every connection when not needed
+     */
+    if (token.is_retry) {
+        *gen_new_token = 1;
+    } else {
+        if (time_diff > ((NEW_TOKEN_LIFETIME * 9) / 10))
+            *gen_new_token = 1;
+    }
+
     ret = 1;
 err:
     cleanup_validation_token(&token);
     OPENSSL_free(remote_addr);
     return ret;
 }
 
+static void generate_new_token(QUIC_CHANNEL *ch, BIO_ADDR *peer)
+{
+    QUIC_CONN_ID rscid = { 0 };
+    QUIC_VALIDATION_TOKEN token;
+    unsigned char buffer[ENCRYPTED_TOKEN_MAX_LEN];
+    unsigned char *ct_buf;
+    size_t ct_len;
+    size_t token_buf_len = 0;
+
+    /* Clients never send a NEW_TOKEN */
+    if (!ch->is_server)
+        return;
+
+    ct_buf = OPENSSL_zalloc(ENCRYPTED_TOKEN_MAX_LEN);
+    if (ct_buf == NULL)
+        return;
+
+    /*
+     * NEW_TOKEN tokens may be used for multiple subsequent connections
+     * within their timeout period, so don't reserve an rscid here
+     * like we do for retry tokens, instead, just fill it with random
+     * data, as we won't use it anyway
+     */
+    rscid.id_len = 8;
+    if (!RAND_bytes_ex(ch->port->engine->libctx, rscid.id, 8, 0)) {
+        OPENSSL_free(ct_buf);
+        return;
+    }
+
+    if (!generate_token(peer, ch->init_dcid, rscid, &token, 0)
+        || !marshal_validation_token(&token, buffer, &token_buf_len)
+        || !encrypt_validation_token(ch->port, buffer, token_buf_len, NULL,
+                                     &ct_len)
+        || ct_len > ENCRYPTED_TOKEN_MAX_LEN
+        || !encrypt_validation_token(ch->port, buffer, token_buf_len, ct_buf,
+                                     &ct_len)
+        || !ossl_assert(ct_len >= QUIC_RETRY_INTEGRITY_TAG_LEN)) {
+        OPENSSL_free(ct_buf);
+        cleanup_validation_token(&token);
```

Comment:
If `generate_token` fails, then `token` will be invalid and `cleanup_validation_token` will access uninitialized memory. So solution is to either add a check for `if token == NULL` within `cleanup_validation_token` (preferred) or only run `cleanup_validation_token` after `generate_token` has successfully completed.

Reply:
The NULL check can't work as written, because the token is allocated on the stack, and always has a non-NULL address `&token`. So the `generate_token` case needs its own separate `if` check, before the rest.

Reply:
Or we could unit with = {0}

Reply:
The first thing that generate_token does is set remote_addr to NULL, which is the only thing that cleanup_validation_token frees at the moment, so we already have the initialization needed.

Reply:
Fair enough, not an immediate problem. Though as a matter of clarity one does not generally expect functions to have done anything useful on error, including initialisation, which is nevertheless good to do, in case some users do rely on side-effects, but I'd still (non-urgently) consider doing any requisite initialisation in the caller, rather than the function that might fail.

Someone might later refactor the function and remove the initialisation on the error path...